feat : multi arch builds and release.sh script

supernovahs opened the pull request:
- Build multi arch image 
- Push performer as multi arch to registry 
- Return aggregator, executor and performer in a mapping for devkit to consume and publish on chain

supernovahs committed:
wip : multi arch builds

supernovahs committed:
build for linux/amd and arm64

supernovahs committed:
fix syntax

supernovahs committed:
oci compliant multi arch manifest  wip

supernovahs committed:
fix script

supernovahs committed:
simplify , add release.sh

supernovahs committed:
use --architectures

supernovahs committed:
update build container

supernovahs committed:
wip build

supernovahs committed:
pass empty context if architectures is not specified

supernovahs committed:
remove artifact id

supernovahs committed:
do not push on build, add release.sh

supernovahs committed:
clean

0xrajath committed:
chore: bump up contracts template dependency

0xrajath committed:
feat: l2 verification changes

0xrajath committed:
chore: update user contracts

supernovahs committed:
add versioning support

supernovahs committed:
move release.sh to .hourglass

supernovahs committed:
add executor and aggregator digest and registry url , and pass a mapping back to devkit

supernovahs committed:
fix build

supernovahs committed:
fix build

supernovahs committed:
aggregator and executor in release

0xrajath committed:
feat: l2 verification contract changes (#64)

**Motivation:**

We need to update the hourglass avs template to support the necessary
changes for l2 verification.

supernovahs committed:
remove artifactId

supernovahs committed:
rename artifacts to artifact

supernovahs committed:
remove semantic versioning , simply take version from devkit

grezle committed:
fix: fetch new ref before checkout

grezle committed:
fix: fetch new ref before checkout (#65)

Make sure we fetch the upgraded contracts ref before checkout to avoid:

```
Updating contracts submodule to desired commit...
fatal: unable to read tree (78749d40126b6fea6977ca90859987c67c1db102)
2025/06/25 23:26:15 upgrade script execution failed: script /var/folders/ph/04f_sgvn4xz7ylqtd00m5tvh0000gn/T/devkit-template-upgrade-1013716629/.devkit/scripts/upgrade exited with code 128
```

This error can be reproduced by creating from `v0.0.12` and upgrading to
`v0.0.13`

supernovahs committed:
push performer in release.sh

supernovahs committed:
performer push in release.sh fix

supernovahs committed:
delete tmp exports  in release.sh

supernovahs committed:
remove duplicate code, use buildcontainer for building local iamge

supernovahs committed:
fix passign image

supernovahs committed:
fix image passing release.sh

grezle committed:
fix: check if .devkit/contracts is inline

supernovahs committed:
add back tmp/info export in build container

supernovahs committed:
take image as arg

supernovahs committed:
Merge branch 'master' into nova/multi-arch-builds

bdchatham commented on .devkit/scripts/build:
```diff
@@ -9,49 +9,62 @@ ensureMake
+# Use provided image name or fallback to default
+if [ -z "$IMAGE_NAME" ]; then
+  project_name=$(basename "$(pwd)")
+  IMAGE_NAME="${project_name}-performer-op-set-1"
```
TODO: can remove the op-set-1 suffix.
TODO: add this suffix even when IMAGE_NAME is passed. If none is passed, fail and return an error.

bdchatham commented on .devkit/scripts/build:
```diff
@@ -9,49 +9,62 @@ ensureMake
-if [ "$REGISTRY" == "null" ] || [ -z "$REGISTRY" ]; then
-    log "Registry Url empty in .hourglass/build.yaml.Passing empty registry url to context. Provide registry url explicitly while calling `devkit avs release publish`"
-    REGISTRY_URL=""
+# Get the build info
```
I think we should avoid storing this information in files, especially since it's outside the devkit/hourglass namespace.

bdchatham commented on .hourglass/build.yaml:
```diff
@@ -1,4 +1,4 @@
 container:
```
TODO: remove this file since we're storing the build/release state in the context.

bdchatham commented on .hourglass/scripts/buildContainer.sh:
```diff
@@ -1,15 +1,68 @@
-tag=$(echo "$buildParams" | yq -r '.container.version')
+# Parse command line arguments
+TAG=""
+REGISTRY=""
```
TODO: we can remove this since we're just building locally.

grezle committed:
fix: check if .devkit/contracts is inline (#66)

Currently, we're checking to see if `.devkit/contracts` is a submodule
by checking if the `.git/modules/...` directory is present, if an
upgrade is attempted and fails and the subsequent fix commit is removed
to get back to a clean slate, then the `.git/modules/...` directory
persists and the check fails essentially bricking the project.

This PR uses git directly to check if the `.devkit/contracts` directory
is registered as a submodule or not, to prevent any reliance on
faulty/flaky state.

bdchatham commented on .hourglass/scripts/buildContainer.sh:
```diff
@@ -1,15 +1,68 @@
-image=$(echo "$buildParams" | yq -r '.container.image')
-tag=$(echo "$buildParams" | yq -r '.container.version')
+# Parse command line arguments
+TAG=""
```
We need to specify this as part of the build.sh inputs. 

TODO: inside the `devkit avs build` command, we should parse the "version" from the context. If none exists, make it 0. During `devkit avs release publish`, increment this value on success.

bdchatham commented on .hourglass/scripts/buildContainer.sh:
```diff
@@ -1,15 +1,68 @@
+IMAGE_ID=$(docker images --format "table {{.ID}}" --no-trunc "$fullImage" | tail -1)
+
+echo "Built container: $fullImage" >&2
+echo "📋 Image ID: $IMAGE_ID" >&2
```
nit: best practice to keep emojis out of stdout.

bdchatham commented on .hourglass/scripts/buildContainer.sh:
```diff
@@ -1,15 +1,68 @@
+echo "📋 Image ID: $IMAGE_ID" >&2
+
+
+# Export build info
```
TODO: we can remove this and instead pass imageId in artifactId. Pass back the imageName as "component" in the context.

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+  exit 1
+fi
+
+# Build temporary image for comparison
+echo "Building temporary image for comparison..." >&2
+temp_build=$(bash -c "$build_cmd --tag release-temp")
+NEW_IMAGE_ID=$(echo "$temp_build" | jq -r '.image_id')
+tempImage=$(echo "$temp_build" | jq -r '.image')
```
TODO: this looks mostly good, but I'd suggest having the tag formatted as `${version}-release`

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+echo "  OperatorSet $executor_operator_set_id (executor): [$executor_json]" >&2
+
+# Build original image using buildContainer.sh
+echo "Building original image..." >&2
```
Seems like we're building the original again. Instead of doing this, we just want to pass the `context.artifact.artifactId` to this script so that the temporary image we create can be compared against that.

supernovahs committed:
add back artifactId

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+echo "Original Image ID: $ORIGINAL_IMAGE_ID" >&2
+echo "New Image ID: $NEW_IMAGE_ID" >&2
+
+# Compare image IDs and set result
```
TODO: if the image changed, the script should fail and inform the user they need to run `devkit avs build` again before performing a release.

supernovahs replied:
https://github.com/Layr-Labs/hourglass-avs-template/pull/63/commits/7cfea6db44c5623949c089f3e8bb67e4dc1752fa

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+# Setup buildx for multi-platform builds
+echo "Setting up multi-platform builder..." >&2
+if ! docker buildx inspect multiarch >/dev/null 2>&1; then
+  docker buildx create --name multiarch --driver docker-container --use >&2
```
I'd add a log statement here stating you are taking this action.

supernovahs replied:
doing it in 106

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+# Build original image using buildContainer.sh
+echo "Building original image..." >&2
+build_cmd=".hourglass/scripts/buildContainer.sh --image $IMAGE" # We are not giving registry url intentionally. Since its not pushing to registry , we just want to compare
+original_build=$(bash -c "$build_cmd")
```
TODO: can remove this and just use artifactId.

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+# Build and push multi-arch
+docker buildx build \
+  --platform linux/amd64,linux/arm64 \
+  --tag "$performer_full_image" \
```
TODO: `ImageName-version`

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+  echo "Error: Could not get performer image digest" >&2
+  exit 1
+fi
+echo "📋 Performer Image Index Digest: $performer_digest" >&2
```
nit: emoji

bdchatham commented on .hourglass/scripts/release.sh:
```diff
@@ -0,0 +1,165 @@
+# Output the operator set mapping to stdout
+echo "$operator_set_mapping_json"
+
+# Return the boolean result (0 = no change, 1 = changed)
```
Can remove

supernovahs committed:
resolutions

supernovahs committed:
do not suffix in release

supernovahs committed:
image-version format

supernovahs committed:
fix buildcontainer

supernovahs committed:
pass correct component image name

supernovahs committed:
fix image name

supernovahs committed:
add log to rebuild and return error if image changed

supernovahs committed:
remove emoji

supernovahs committed:
remove unneeded code

supernovahs committed:
remove storing in files

supernovahs committed:
capture build cotnainer output correctly

supernovahs committed:
fix build

supernovahs committed:
move to image:version

supernovahs committed:
Merge branch 'master' into nova/multi-arch-builds

supernovahs committed:
Merge branch 'master' into nova/multi-arch-builds

bdchatham reviewed (commented):
Generally looks good! Going to take another pass through with fresh eyes in the morning. Local testing also working well.

supernovahs committed:
registry url => registry

supernovahs committed:
fix release.sh
